core/src/cluster_nodes.rs

```diff
@@ -211,6 +229,67 @@ fn get_nodes(cluster_info: &ClusterInfo, stakes: &HashMap<Pubkey, u64>) -> Vec<N
     .collect()
 }
 
+impl<T> ClusterNodesCache<T> {
+    pub(crate) fn new(
+        // Capacity of underlying LRU-cache in terms of number of epochs.
+        cap: usize,
+        // A time-to-live eviction policy is enforced to refresh entries in
+        // case gossip contact-infos are updated.
+        ttl: Duration,
+    ) -> Self {
+        Self {
+            cache: Mutex::new(LruCache::new(cap)),
+            ttl,
+        }
+    }
+}
+
+impl<T: 'static> ClusterNodesCache<T> {
+    fn get_cache_entry(&self, epoch: Epoch) -> Arc<Mutex<CacheEntry<T>>> {
+        let mut cache = self.cache.lock().unwrap();
+        match cache.get(&epoch) {
+            Some(entry) => Arc::clone(entry),
+            None => {
+                let entry = Arc::default();
+                cache.put(epoch, Arc::clone(&entry));
+                entry
+            }
+        }
+    }
+
+    pub(crate) fn get(
+        &self,
+        shred_slot: Slot,
+        root_bank: &Bank,
+        cluster_info: &ClusterInfo,
+    ) -> Arc<ClusterNodes<T>> {
+        let epoch = root_bank.get_leader_schedule_epoch(shred_slot);
+        let entry = self.get_cache_entry(epoch);
+        // Hold the lock on the entry here so that, if needed, only
+        // one thread recomputes cluster-nodes for this epoch.
+        let mut entry = entry.lock().unwrap();
+        if let Some((asof, nodes)) = entry.deref() {
+            if asof.elapsed() < self.ttl {
```

Comment:
If we enforce: https://github.com/solana-labs/solana/pull/18971/files#r681451853 the `ttl` doesn't seem necessary since the set of epoch staked nodes for the known epochs <  `root_bank.get_leader_schedule_epoch(root_slot)` should always return the same result

Reply:
Ah shucks, I misread this, I thought we were caching the `epoch_staked_nodes` here: https://github.com/solana-labs/solana/pull/18971/files#diff-ef69c0f93244cfe4b45f4d6ca5af4e91b59c123ef228cb0e73ac80a21c078eafR276 which never changes. 

Does it makes sense to cache the `epoch_staked_ndoes` as well so we can reuse it when reconstructing the cache entry based on new gossip contact-infos? i.e. changing the `CacheEntry<T>` to be `type CacheEntry<T> = Option<(/*as of:*/ Instant, /*epoch staked nodes*/ HashMap<Pubkey, u64>, Arc<ClusterNodes<T>>)>;`

Reply:
https://github.com/solana-labs/solana/pull/19090 will implement copy-on-write for staked-nodes and so wouldn't there be much room left here to optimize. Even without that commit I doubt that it will make much of difference. Leaning on keeping things simple for now.

Reply:
Sounds good, thanks!